device/core/src/module_client.ts

```diff
@@ -126,11 +131,76 @@ export class ModuleClient extends InternalClient {
+   *
+   * @param deviceId      target device identifier
+   * @param moduleId      target module identifier on the device identified with the `deviceId` argument
+   * @param methodParams  parameters of the direct method call
+   * @param callback      callback that will be invoked either with an Error object or the result of the method call.
+   */
+  invokeMethod(deviceId: string, methodParams: MethodParams, callback: MethodCallback): void;
+  invokeMethod(deviceId: string, moduleId: string, methodParams: MethodParams, callback: MethodCallback): void;
+  invokeMethod(deviceId: string, moduleIdOrMethodParams: string | MethodParams, methodParamsOrCallback: MethodParams | MethodCallback, callback?: MethodCallback): void {
+    /*Codes_SRS_NODE_MODULE_CLIENT_16_093: [`invokeMethod` shall throw a `ReferenceError` if the `deviceId` argument is falsy.]*/
+    if (!deviceId) {
+      throw new ReferenceError('deviceId cannot be \'' + deviceId + '\'');
+    }
+    /*Codes_SRS_NODE_MODULE_CLIENT_16_094: [`invokeMethod` shall throw a `ReferenceError` if the `moduleIdOrMethodParams` argument is falsy.]*/
+    if (!moduleIdOrMethodParams) {
+      if (typeof methodParamsOrCallback === 'object') {
+        throw new ReferenceError('moduleId cannot be \'' + moduleIdOrMethodParams + '\'');
+      } else {
+        throw new ReferenceError('methodParams cannot be \'' + moduleIdOrMethodParams + '\'');
+      }
+    }
+
+    /*Codes_SRS_NODE_MODULE_CLIENT_16_095: [`invokeMethod` shall throw a `ReferenceError` if the `deviceId` and `moduleIdOrMethodParams` are strings and the `methodParamsOrCallback` argument is falsy.]*/
+    if (deviceId && typeof deviceId === 'string'
+        && moduleIdOrMethodParams && typeof moduleIdOrMethodParams === 'string'
+        && !methodParamsOrCallback) {
+      throw new ReferenceError('methodParams cannot be \'' + methodParamsOrCallback + '\'');
+    }
+
+    const actualModuleId     = typeof moduleIdOrMethodParams === 'string' ? moduleIdOrMethodParams : null;
+    const actualMethodParams = typeof moduleIdOrMethodParams === 'object' ? moduleIdOrMethodParams : methodParamsOrCallback;
+    const actualCallback     = typeof methodParamsOrCallback === 'function' ? methodParamsOrCallback : callback;
+
+
+    /*Codes_SRS_NODE_MODULE_CLIENT_16_096: [`invokeMethod` shall throw a `ArgumentError` if the `methodName` property of the `DeviceMethodParams` argument is falsy.]*/
+    if (!(actualMethodParams as MethodParams).methodName) {
+      throw new errors.ArgumentError('the name property of the methodParams argument cannot be \'' + (actualMethodParams as MethodParams).methodName + '\'');
+    }
+
+    /*Codes_SRS_NODE_MODULE_CLIENT_16_097: [`invokeMethod` shall call the `invokeMethod` API of the `DeviceMethodClient` API that was created for the `ModuleClient` instance.]*/
+    this._methodClient.invokeMethod(deviceId, actualModuleId, actualMethodParams as MethodParams, actualCallback);
```

Comment:
>_methodClient [](start = 9, length = 13)

If _methodClient is null, presumably because this is a moduleClient talking to IoTHub, we should fail earlier with a more meaningful error.

Reply:
should never be null. if it were it'd be a bug which is why i'm fine not testing it.

---
In reply to: [197227257](https://github.com/Azure/azure-iot-sdk-node/pull/322#discussion_r197227257) [](ancestors = 197227257)